apollo-federation/src/sources/connect/models/validation.rs

```diff
@@ -49,3 +39,225 @@
 - Missing $this
 
 */
+
+use apollo_compiler::schema::{Component, Directive};
+use apollo_compiler::Schema;
+use itertools::Itertools;
+use std::fmt::Display;
+use url::Url;
+
+/// Validate the connectors-related directives `@source` and `@connect`.
+///
+/// This function attempts to collect as many validation errors as possible, so it does not bail
+/// out as soon as it encounters one.
+pub fn validate(schema: Schema) -> Vec<ValidationError> {
+    let mut source_names = Vec::new();
+    let errors = schema
+        .schema_definition
+        .directives
+        .iter()
+        .filter_map(|directive| {
```

Comment:
Nit: I think that it might be more clear to have a `filter` followed by a `map` here to help break apart the nesting.

Reply:
Right now `=="source"` _looks_ like a filter, but there's going to be a `=="connect"` branch as well—it's really dispatching. That said, like the side-effects discussion, I think we'll best know how to clean this up once most of the top-level logic is in.